include/dslash.h

```diff
@@ -77,9 +78,36 @@ namespace quda
       strncat(aux[kernel_type], aux_base, TuneKey::aux_n - 1);
     }
 
-    virtual bool tuneGridDim() const { return false; }
+    virtual bool tuneGridDim() const { return arg.kernel_type == EXTERIOR_KERNEL_ALL && arg.shmem > 0; }
     virtual unsigned int minThreads() const { return arg.threads; }
 
+    virtual unsigned int minGridSize() const
+    {
+      /* when using nvshmem we perform the exterior Dslash using a grid strided loop and uniquely assign communication
+       * directions to CUDA block and have all communication directions resident. We therefore figure out the number of
+       * communicating dimensions and make sure that the number of blocks is a multiple of the communicating directions (2*dim)
+       */
+      if (arg.kernel_type == EXTERIOR_KERNEL_ALL && arg.shmem > 0) {
+        int nDimComms = 0;
+        for (int d = 0; d < in.Ndim(); d++) nDimComms += arg.commDim[d];
+        return ((deviceProp.multiProcessorCount) / (2 * nDimComms)) * (2 * nDimComms);
```

Comment:
Sanity check: you can never hit this codepath when `nDimComms` is zero, correct? Is this still true with Schwarz preconditioners/the 3-d laplace op (for covariant derivatives)? @maddyscientist

Reply:
This won't be a problem for a Schwarz preconditioner since we never have exterior kernels for that.  It will be fine also for 3-d Laplace since `commDim` will be set appropriately.